evm/src/main/java/org/hyperledger/besu/evm/UInt256.java

```diff
@@ -0,0 +1,413 @@
+   * @return 0 if a == b, negative if a &lt; b and positive if a &gt; b.
+   */
+  public static int compare(final UInt256 a, final UInt256 b) {
+    int comp = Integer.compare(a.length, b.length);
+    if (comp != 0) return comp;
+    for (int i = a.length - 1; i >= 0; i--) {
+      comp = Integer.compareUnsigned(a.limbs[i], b.limbs[i]);
+      if (comp != 0) return comp;
+    }
+    return 0;
+  }
+
+  private int[] shiftLeftWithCarry(final int shift) {
+    int limbShift = shift / 32;
+    int bitShift = shift % 32;
+    if (bitShift == 0) {
+      return Arrays.copyOfRange(limbs, limbShift, length + limbShift + 1);
+    }
+
+    int[] res = new int[length + limbShift + 1];
+    int j = limbShift;
+    int carry = 0;
+    for (int i = 0; i < length; ++i, ++j) {
+      res[j] = (limbs[i] << bitShift) | carry;
+      carry = limbs[i] >>> (32 - bitShift);
+    }
+    res[j] = carry; // last carry
+    return res;
+  }
+
+  /**
+   * Shifts value to the left.
+   *
+   * @param shift number of bits to shift. If negative, shift right instead.
+   * @return Shifted UInt256 value.
+   */
+  public UInt256 shiftLeft(final int shift) {
+    if (shift < 0) return shiftRight(-shift);
+    if (shift == 0 || isZero()) return this;
+    if (shift >= length * 32) return ZERO;
+    return new UInt256(shiftLeftWithCarry(shift));
+  }
+
+  /**
+   * Shifts value to the right.
+   *
+   * @param shift number of bits to shift. If negative, shift left instead.
+   * @return Shifted UInt256 value.
+   */
+  public UInt256 shiftRight(final int shift) {
+    if (shift < 0) return shiftLeft(-shift);
+    if (shift == 0 || isZero()) return this;
+    if (shift >= length * 32) return ZERO;
+
+    int limbShift = shift / 32;
+    int bitShift = shift % 32;
+    if (bitShift == 0) {
+      return new UInt256(Arrays.copyOfRange(limbs, limbShift, length), length - limbShift);
+    }
+
+    int[] res = new int[this.limbs.length - limbShift];
+    int j = this.length - 1 - limbShift; // res index
+    int carry = 0;
+    for (int i = this.length - 1; j >= 0; i--, j--) {
+      int r = (limbs[i] >>> bitShift) | carry;
+      res[j] = r;
+      carry = limbs[i] << (32 - bitShift);
+    }
+    return new UInt256(res);
+  }
+
+  /**
+   * Reduce modulo divisor.
+   *
+   * @param divisor The modulus of the reduction
+   * @return The remainder modulo {@code divisor}.
+   */
+  public UInt256 mod(final UInt256 divisor) {
+    if (divisor.isZero()) return ZERO;
+    int cmp = compare(this, divisor);
+    if (cmp < 0) return this;
+    if (cmp == 0) return ZERO;
+
+    int n = divisor.length;
+    int m = this.length - n;
+    if (n == 1) {
+      long d = divisor.limbs[0] & 0xFFFFFFFFL;
+      long rem = 0;
+      // Process from most significant limb downwards
+      for (int i = length - 1; i >= 0; i--) {
+        long cur = (rem << 32) | (limbs[i] & 0xFFFFFFFFL);
+        rem = Long.remainderUnsigned(cur, d);
+      }
+      return new UInt256(new int[] {(int) rem});
+    }
```

Comment:
Another optimization I think you can do here is that if `this.length == 1` you can just do this[0] % divisor[0]

Reply:
Yes thanks, will add this one